Signal/src/ViewControllers/LoadingViewController.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import PromiseKit
+
+// The initial presentation is intended to be indistinguishable from the Launch Screen.
+// After a delay we present some "loading" UI so the user doesn't think the app is frozen.
+@objc
+public class LoadingViewController: UIViewController {
+
+    var logoView: UIImageView!
+    var topLabel: UILabel!
+    var bottomLabel: UILabel!
+
+    override public func loadView() {
+        self.view = UIView()
+        view.backgroundColor = UIColor.ows_materialBlue
+
+        self.logoView = UIImageView(image: #imageLiteral(resourceName: "logoSignal"))
+        view.addSubview(logoView)
+
+        logoView.autoCenterInSuperview()
+        logoView.autoPinToSquareAspectRatio()
+        logoView.autoMatch(.width, to: .width, of: view, withMultiplier: 1/3)
+
+        self.topLabel = buildLabel()
+        topLabel.alpha = 0
+        topLabel.font = UIFont.ows_dynamicTypeTitle2
+        topLabel.text = NSLocalizedString("DATABASE_VIEW_OVERLAY_TITLE", comment: "Title shown while the app is updating its database.")
+
+        self.bottomLabel = buildLabel()
+        bottomLabel.alpha = 0
+        bottomLabel.font = UIFont.ows_dynamicTypeBody
+        bottomLabel.text = NSLocalizedString("DATABASE_VIEW_OVERLAY_SUBTITLE", comment: "Subtitle shown while the app is updating its database.")
+
+        let labelStack = UIStackView(arrangedSubviews: [topLabel, bottomLabel])
+        labelStack.axis = .vertical
+        labelStack.alignment = .center
+        labelStack.spacing = 8
+        view.addSubview(labelStack)
+
+        labelStack.autoPinEdge(.top, to: .bottom, of: logoView, withOffset: 20)
+        labelStack.autoPinLeadingToSuperviewMargin()
+        labelStack.autoPinTrailingToSuperviewMargin()
+        labelStack.setCompressionResistanceHigh()
+        labelStack.setContentHuggingHigh()
+    }
+
+    var isShowingTopLabel = false
+    var isShowingBottomLabel = false
+    override public func viewDidAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        // We only show the "loading" UI if it's a slow launch. Otherwise this ViewController
+        // should be indistinguishable from the launch screen.
+        let kTopLabelThreshold: TimeInterval = 5
+        DispatchQueue.main.asyncAfter(deadline: .now() + kTopLabelThreshold) { [weak self] in
```

Comment:
Ideally, main thread should not be blocked during a long db migration or when we rebuild database extensions.  But it could. Please add to the test plan some scenario where we rebuild all db extensions and verify that this works correctly.

Reply:
Sure thing.